Add missing config and initial staking

iamrz1 opened the pull request:
## Fixes
https://github.com/0chain/0chain/issues/1700
https://github.com/0chain/0chain/issues/1707

## Changes

## Need to be mentioned in CHANGELOG.md?

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

iamrz1 committed:
Added missing values in /storage_config

iamrz1 committed:
Added initial staking

peterlimg commented on code/go/0chain.net/chaincore/chain/entity.go:
```diff
@@ -611,6 +618,69 @@ func (c *Chain) setupInitialState(initStates *state.InitStates) util.MerklePatri
+		sp := stakepool.StakePool{}
+		if err := mpt.GetNodeValue(util.Path(encryption.Hash(key)), &sp); err != nil {
+			if err != util.ErrValueNotPresent {
+				continue
```
Please remember that 'node not found' error should always be reported, so let's return error instead of ignoring it.

iamrz1 replied:
I think the context here is a bit different. stakePool is with a certain key can exist or it can be a new stakepool to be inserted, both are expected in this case (since we are dealing with initial stake pools). The error that I'm skipping is anything except `node not found`, because i'm not sure whether I should report it back as an error, or just print as log. I'm open to suggestion about it.

peterlimg replied:
Yes, it's a bit different, but still we need to panic here instead of continue or logging only, it is one of the most dangerous behaviors to skip a 'node not found' error. So please do 
```go
if state.ErrInvalidState(err) {
  logging.Logger.Panic("could not initialize staking", zap.Err(err))
}
```

peterlimg commented on code/go/0chain.net/chaincore/chain/entity.go:
```diff
@@ -611,6 +618,69 @@ func (c *Chain) setupInitialState(initStates *state.InitStates) util.MerklePatri
+		sp.Pools[v.ClientID] = &stakepool.DelegatePool{
+			Balance:      v.Tokens,
+			Reward:       0,
+			Status:       0,
```
We do not need to set zero values explicitly

iamrz1 replied:
Sure, I can remove them.

peterlimg commented on code/go/0chain.net/chaincore/chain/entity.go:
```diff
@@ -611,6 +618,69 @@ func (c *Chain) setupInitialState(initStates *state.InitStates) util.MerklePatri
+			RoundCreated: 0,
+		})
+
+		if _, err := mpt.Insert(util.Path(encryption.Hash(key)), &sp); err != nil {
```
Please call  `sp.Save()` to save the stake pool

iamrz1 replied:
I am trying to use `sp.Save()`, but can't find a suitable value for this parameter `balances cstate.StateContextI`. Apparently at this point in code `c.GetStateContextI()` returns nil value and causes panic.

peterlimg replied:
See code in line 581, we can pass in the `stateCtx`

iamrz1 replied:
Great! Will do.

peterlimg commented on code/go/0chain.net/smartcontract/stakepool/spenum/enums.go:
```diff
@@ -22,6 +22,15 @@ func (p Provider) String() string {
 	return providerString[p]
 }
 
+func ToProviderID(ps string) Provider {
```
It's provider type,  so should be `ToProviderType`

iamrz1 replied:
Sure.

iamrz1 committed:
Changed mpt calls to stakePool specific methods

peterlimg commented on code/go/0chain.net/chaincore/chain/entity.go:
```diff
@@ -618,43 +618,49 @@ func (c *Chain) setupInitialState(initStates *state.InitStates) util.MerklePatri
-		if err := mpt.GetNodeValue(util.Path(encryption.Hash(key)), &sp); err != nil {
-			if err != util.ErrValueNotPresent {
-				continue
+		if knownStakePoolKeys[stakePoolKey] {
```
We don't need to create a map to check if the stakePool exist, checking the MPT by calling `sp.Get()` is enough, we won't re-enter this function later after we refactor it to only run genesis block once. So let's keep the changes simple, and avoid exposing the `stakepool.StakePoolKey()` function as well

iamrz1 replied:
Got it. One of the reasons to add the map was to add up the tokens for stakes with same clientID and and provider. It would also reduce calls to mpt for each entry in stakepool. But if that's fine, I'll  remove it.

peterlimg replied:
We don't need to add up the tokens for stakes, we would only initialize stakes **once** on genesis block, which is round 1. It's a deterministic result, and we shall never change the staked value on genesis block later again. So please remove the `add up token` code and only do existence checking.

iamrz1 replied:
okay, so removing code to support adding up stakes for cases as below:
```
  - provider_id: x
    provider_type: blobber
    client_id: y
    tokens: 50000000000

  - provider_id: x
    provider_type: blobber
    client_id: y
    tokens: 10000000000
```
resulting in tokens = 60000000000

iamrz1 committed:
Removed knownStakePoolKeys map when applying intial stakes

iamrz1 committed:
Removed adding up tokens for duplicate intial stake entries

peterlimg committed:
Do not return error if stake exist

Before we implemented code to stop genesis block re-entering, genesis block generationg code will be called everytime when node starts, so staking would aleady exist, and we need to continue instead of returning to cause panic.

peterlimg committed:
Merge branch 'staging' into 0chain-1755

* staging: (32 commits)
  fix: remove check for lfb
  fix: check for round/lfb before genesis setup
  make rm and wm hash different
  make rm and wm hash different
  make rm and wm hash different
  make rm and wm hash different
  merge PR1873
  fix
  fix wirtemarker tx hash
  touch
  touch
  lock network until all nodes are ready on miner fault tolerance test
  remove commented code
  lock network until all nodes are registered in fault tolerance test
  touch
  add type to fee reward
  remove debug line
  reduce bank data list length
  reduce bank data list length
  reduce bank data list length
  ...

peterlimg committed:
Return error for duplicate staking as genesis block re-enter issue is fixed

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | 0chain-1755  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |
